src/utils/lbann_library.cpp

```diff
@@ -48,193 +48,189 @@ std::shared_ptr<thread_pool> construct_io_thread_pool(lbann_comm *comm) {
-    if (opts->has_string("share_testing_data_readers")) {
-      is_shared_testing_data_reader = opts->get_bool("share_testing_data_readers");
-    }
-    init_data_readers(comm, pb, data_readers, is_shared_training_data_reader, is_shared_testing_data_reader);
-
-    // hack to prevent all data readers from loading identical data; instead,
-    // share a single copy. See data_reader_jag_conduit_hdf5 for example
-    if (first_model) {
-      if (opts->has_string("share_data_reader_data")) {
-        for (auto t : data_readers) {
-          opts->set_ptr((void*)t.second);
-        }
+  // Initialize data readers
+  //@todo: code not in place for correctly handling image preprocessing
+  std::map<execution_mode, generic_data_reader *> data_readers;
+  bool is_shared_training_data_reader = pb_model->shareable_training_data_reader();
+  bool is_shared_testing_data_reader = pb_model->shareable_testing_data_reader();
+  if (opts->has_string("share_testing_data_readers")) {
+    is_shared_testing_data_reader = opts->get_bool("share_testing_data_readers");
+  }
+  init_data_readers(comm, pb, data_readers, is_shared_training_data_reader, is_shared_testing_data_reader);
+
+  // hack to prevent all data readers from loading identical data; instead,
+  // share a single copy. See data_reader_jag_conduit_hdf5 for example
+  if (first_model) {
+    if (opts->has_string("share_data_reader_data")) {
+      for (auto t : data_readers) {
+        opts->set_ptr((void*)t.second);
       }
     }
+  }
 
-    // User feedback
-    print_parameters(comm, pb);
+  // User feedback
+  print_parameters(comm, pb);
 
-    // Initalize model
-    model = proto::construct_model(comm,
-                                   data_readers,
-                                   pb.optimizer(),
-                                   pb.model());
-    model->setup(io_thread_pool);
+  // Initalize model
+  std::unique_ptr<model> ret_model{
+    proto::construct_model(comm,
+                           data_readers,
+                           pb.optimizer(),
+                           pb.model())
+  };
+  ret_model->setup(std::move(io_thread_pool));
 
-    if(opts->get_bool("disable_background_io_activity")) {
-      model->allow_background_io_activity(false);
-    }
+  if(opts->get_bool("disable_background_io_activity")) {
+    ret_model->allow_background_io_activity(false);
+  }
 
-    //under development; experimental
-    if (opts->has_bool("use_data_store") && opts->get_bool("use_data_store")) {
-      if (master) {
-        std::cout << "\nUSING DATA STORE!\n\n";
-      }
-      for (auto r : data_readers) {
-        if (!r.second) continue;
-        r.second->setup_data_store(model);
-      }
+  //under development; experimental
+  if (opts->has_bool("use_data_store") && opts->get_bool("use_data_store")) {
+    if (master) {
+      std::cout << "\nUSING DATA STORE!\n\n";
     }
-
-    if (opts->has_string("create_tarball")) {
-      finalize(comm);
-      return 0;
```

Comment:
This code block was just wrong and if it were ever used, it would, with high probability, result in a segfault. In the single-model case, one would probably call `model_ptr->anything()` on a null `model_ptr` and in the multimodel case, one's next call to this function would attempt to operate on a finalized comm. So, question: What is the intent here? What behavior should happen in the multi-model case? I'm guessing the right behavior is something like:
```cpp
int main(int argc, char** argv)
{
    create_array_of_models();
    if (opts->has_string("create_tarball")) {
        return EXIT_SUCCESS;
    }
    train_models();
}```
But I wasn't 100% sure so I only added code like this to `lbann.cpp`. Please advise on what to do for other front-ends.